crmsh/report/core.py

```diff
@@ -2,427 +2,489 @@
-    env_dict["EXTRA_LOGS"] = constants.EXTRA_LOGS
-    env_dict["PCMK_LOG"] = constants.PCMK_LOG
-    env_dict["VERBOSITY"] = int(config.report.verbosity) or (1 if config.core.debug else 0)
-
-    res_str = ""
-    for k, v in env_dict.items():
-        res_str += " {}={}".format(k, v)
-    return res_str
-
-def get_log():
-    """
-    get the right part of the log
-    """
-    outf = os.path.join(constants.WORKDIR, constants.HALOG_F)
-
-    # collect journal from systemd unless -M was passed
-    if constants.EXTRA_LOGS:
-        utillib.collect_journal(constants.FROM_TIME,
-                                constants.TO_TIME,
-                                os.path.join(constants.WORKDIR, constants.JOURNAL_F))
-
-    if constants.HA_LOG and not os.path.isfile(constants.HA_LOG):
-        if not is_collector(): # warning if not on slave
-            logger.warning("%s not found; we will try to find log ourselves", constants.HA_LOG)
-            constants.HA_LOG = ""
-    if not constants.HA_LOG:
-        constants.HA_LOG = utillib.find_log()
-    if (not constants.HA_LOG) or (not os.path.isfile(constants.HA_LOG)):
-        if constants.CTS:
-            pass  # TODO
-        else:
-            logger.warning("not log at %s", constants.WE)
-        return
+        p.join()
+
+
+def start_collector(node: str, context: Context) -> None:
+    """
+    Start collector at specific node
+    """
+    cmd = f"{constants.BIN_COLLECTOR} '{context}'"
+    err = ""
 
-    if constants.CTS:
-        pass  # TODO
+    if node == context.me:
+        code, out, err = ShellUtils().get_stdout_stderr(cmd)
     else:
-        try:
-            getstampproc = utillib.find_getstampproc(constants.HA_LOG)
-        except PermissionError:
-            return
-        if getstampproc:
-            constants.GET_STAMP_FUNC = getstampproc
-            if utillib.dump_logset(constants.HA_LOG, constants.FROM_TIME, constants.TO_TIME, outf):
-                utillib.log_size(constants.HA_LOG, outf+'.info')
+        node = f"{context.ssh_user}@{node}" if context.ssh_user else node
+        cmd = cmd.replace('"', '\\"')
+        cmd = f'ssh {constants.SSH_OPTS} {node} "{context.sudo} {cmd}"'
+        code, out, err = sh.LocalShell().get_rc_stdout_stderr(context.ssh_user, cmd)
+
+    if code != 0:
+        logger.warning(err)
+    # ERROR/WARNING/DEBUG messages
+    if err:
+        print(err, file=sys.stderr)
+    if out == '':
+        return
+
+    compress_data = ""
+    for data in out.split("\n"):
+        if data.startswith(constants.COMPRESS_DATA_FLAG):
+            # crm report data from collector
+            compress_data = data.lstrip(constants.COMPRESS_DATA_FLAG)
         else:
-            logger.warning("could not figure out the log format of %s", constants.HA_LOG)
+            # log data from collector
+            print(data)
 
+    cmd = f"(cd {context.work_dir} && tar xf -)"
+    ShellUtils().get_stdout(cmd, input_s=eval(compress_data))
```

Comment:
What is the use of `eval`? The data format seems to be complex here. Can you add some comments to describe the format?

Reply:
It can be just written in binary.

```python
sys.stdout.write(constants.COMPRESS_DATA_FLAG)
subprocess.run(cmd, stdout=sys.stdout)
```

or

```python
_, out, err = ShellUtils().get_stdout_stderr(cmd, raw=True)
sys.stdout.write(constants.COMPRESS_DATA_FLAG)
sys.stdout.buffer.write(out)
```

or even

```python
_, out, err = ShellUtils().get_stdout_stderr(cmd, raw=True)
sys.stdout.write(constants.COMPRESS_DATA_FLAG)
sys.stdout.write(out.encode('latin-1'))
```

Reply:
Not work for me for the above ways, please submit commits after this PR gets merged